[quant][graphmode] Invoke more passes in `insertObservers`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30476 [quant][graphmode] Add tests for quantizing traced models
* #30475 Refactor test_quantization.py and enable `test_nested`
* #30474 [quant][graphmode] Refactor bias and weight check and add aten::linear pattern
* **#30473 [quant][graphmode] Invoke more passes in `insertObservers`**

Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795722](https://our.internmc.facebook.com/intern/diff/D18795722)

jerryzh168 committed:
[quant][graphmode] Invoke more passes in `insertObservers`

Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

lly-zero-one commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -365,8 +366,12 @@ void InsertObserversHelper::insertObservers(
   auto graph = method.graph();
   ConstantPropagation(graph);
+  // To cleanup traced graph
+  ConstantPooling(graph);
```
Just curious, is there any order we can follow for the optimization passes?

jerryzh168 replied:
not sure, looks like in graph_executor constant pooling is done before constant propagation, I'll change the order

ZolotukhinM replied:
If we want to cleanup traced graphs, shouldn't we do it in tracing?

jerryzh168 replied:
Not sure why we don't do this in traced graph, cc @jamesr66a

jamesr66a replied:
Was speaking with @suo yesterday. It's a pretty common request to add more "clean-up" passes to Module emission (i.e. tracing or scripting) as opposed to just in the GraphExecutor. I think this PR is good, but it would also be valid to just add these things to tracing

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -365,8 +366,12 @@ void InsertObserversHelper::insertObservers(
+  ConstantPooling(graph);
   // must do constant propagation first before replacement
   replaceConvolutionWithConv2d(graph);
+  // fuse exploded linear into aten::linear
```
What do you mean by exploded?

jerryzh168 replied:
exploded means: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization_patterns.h#L25-L64

jamesr66a replied:
I think "decomposed" is the more common term in the JIT code. e.g. `decompose_ops.cpp`

jerryzh168 replied:
OK

jerryzh168 committed:
Update on "[quant][graphmode] Invoke more passes in `insertObservers`"


Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Invoke more passes in `insertObservers`"


Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Invoke more passes in `insertObservers`"


Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795722](https://our.internmc.facebook.com/intern/diff/D18795722)

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -364,9 +365,13 @@ void InsertObserversHelper::insertObservers(
   // must do constant propagation first before replacement
   replaceConvolutionWithConv2d(graph);
+  // fuse decomposed linear into aten::linear
+  FuseLinear(graph);
```
Though for this pass I remember there was some back-and-forth earlier about how it's not profitable in some cases, and might mess up some downstream systems (e.g. CUDA fuser). We could add it generally and see if anything breaks

jerryzh168 replied:
yeah we can't do this by default since it doesn't work for autodiff, but @wanchaol is working on this.

jerryzh168 committed:
Update on "[quant][graphmode] Invoke more passes in `insertObservers`"


Summary:
Invoked `ConstantPooling` and `FuseLinear` pass before
`insertObservers`.
`ConstantPooling` is for cleanning up traced graph, e.g. when we
have to constant node that has the same value, this pass will merge them,
this allows us to have less quantization patterns
`FuseLinear` is to merge the exploded linear function into `aten::linear` so
that we can quantize this function properly. We need to fuse it because right now
the way we recognize weight and bias is by matching the argument position in certain function
calls, e.g. 1st argument of aten::conv2d is weight. Therefore we have to preserve
the bounary of the linear function to recognize the weight of linear. Since in the exploded
linear code, input of addmm is transposed weight rather than the original weight of linear.

Test Plan:
This is needed for quantizing traced model tests to pass
Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18795722](https://our.internmc.facebook.com/intern/diff/D18795722)

[ghstack-poisoned]
